WIP: More warnings as errors

TomAtkinsonArm opened the pull request:
## Description

As discussed on the call this PR will be used to determine what error level we want to conform to and the work needed to conform to said level

## General Checklist:

Please ensure the following points are checked:

- [ ] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [ ] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [ ] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [ ] I have commented any added functions (in line with Doxygen)
- [ ] I have commented any code that could be hard to understand
- [ ] My changes do not add any new compiler warnings
- [ ] My changes do not add any new validation layer errors or warnings
- [ ] I have used existing framework/helper functions where possible
- [ ] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [ ] My changes do not add any regressions
- [ ] I have tested every sample to ensure everything runs correctly
- [ ] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [ ] I have tested the sample on at least one compliant Vulkan implementation
- [ ] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [ ] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [ ] Any dependent assets have been merged and published in downstream modules
- [ ] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

TomAtkinsonArm committed:
More warnings as errors

gpx1000 commented on framework/CMakeLists.txt:
```diff
@@ -402,9 +402,9 @@ endif()
+        target_compile_options(${PROJECT_NAME} PRIVATE -Werror -Wall)
     elseif ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
-        target_compile_options(${PROJECT_NAME} PRIVATE /W3 /WX)
+        target_compile_options(${PROJECT_NAME} PRIVATE /Wall /WX)
```
Been a bit since last I checked on what compile flags MSVC supports, However, I thought Wall is clang/gcc only and msvc would use /W4

As there's going to be several warnings, would you like some help on the cleanup?  We can all use this branch after we get the plugins PR merged.

SaschaWillems replied:
MSVC does indeed support /Wall. I'm not sure if we actually want to use that. Details on this can be found here: https://docs.microsoft.com/en-us/cpp/preprocessor/compiler-warnings-that-are-off-by-default?view=msvc-160

If we enable this for MSVC we'll probably need to explicitly disable different warnings, esp. with VS2019.

gpx1000 replied:
The trick we'll need to use is to specify that the compile options are private the the framework and samples so they don't also go into the dependencies.  That should cut down quite a bit.

Looking locally I'm seeing a fair number that can be auto resolved (clion has these tools... ;-)).  Mostly static analysis tools can get rid of very simple warnings.  However, I'm more concerned with the false positives.  We'll need to target very specific parts of the code with a macro that disables specific warnings as there will be several instances of unused variable that gets set in C++ but used in shaders.

SaschaWillems commented:
Just did a quick run with /Wall on Visual Studio 2019, which results in 1727 warnings in the framework and failing the build due to an error in the Vulkan headers (!) in vulkan_core.h on line 2056. So we also probably need to exclude certain files from either warnings at all or /Wall.

gpx1000 committed:
remove warnings from pr-320

gpx1000 committed:
switch back to generic warn macro removal.

TomAtkinsonArm committed:
Merge pull request #7 from gpx1000/warning_removal_pr-320

Warning removal pr 320

TomAtkinsonArm commented:
Hey @gpx1000,

@cforfang had a quick look over the initial fixes and spotted our use of `LOG*()`

Currently in `common/logging.h` we end the definitions with ```;``` which causes the error when using the LOG definitions elsewhere. It would be better to continue to end our log calls with the semi colon and remove the semi colons from the definitions below.

```
#define LOGI(...) spdlog::info(__VA_ARGS__);
#define LOGW(...) spdlog::warn(__VA_ARGS__);
#define LOGE(...) spdlog::error("[{}:{}] {}", __FILENAME__, __LINE__, fmt::format(__VA_ARGS__));
#define LOGD(...) spdlog::debug(__VA_ARGS__);
```

If your in agreement I will do a quick find and replace to revert this specific change from your contribution.

gpx1000 commented:
Hey @TomAtkinsonArm and @cforfang 

yep, that is a good idea.  I personally find it easier to read code when as a rule macros behave like regular function calls so I typically require similar solutions in my personal code (no ';' allowed to terminate a macro function).

TomAtkinsonArm commented:
@gpx1000 thanks for your input, Ill change that in my latest commit

TomAtkinsonArm commented:
Reverted the LOG stuff, combined some third parties into a single header (glm, maybe vulkan headers? not decided yet)

```
#	pragma warning(disable : 4820 4061 4625 4626 4355 5204 5220 5026 5027 4100 4324 4514 4365 5045 4263 4264 4389 4868 4623 4242 4244 5039)

// C4820 - 'bytes' bytes padding added after construct 'member_name'
// C4061 - enumerator 'identifier' in switch of enum 'enumeration' is not explicitly handled by a case label
// C4625 - 'derived class' : copy constructor was implicitly defined as deleted because a base class copy constructor is inaccessible or deleted
// C4626 - 'derived class' : assignment operator was implicitly defined as deleted because a base class assignment operator is inaccessible or deleted
// C4355 - 'this' : used in base member initializer list
// C5026 - 'type': move constructor was implicitly defined as deleted
// C5027 - 'type': move assignment operator was implicitly defined as deleted
// C4100 - unreferenced formal parameter
// C4324 - structure was padded due to alignment specifier
// C4514 - unreferenced inline function has been removed
// C4365 - conversion from x to y, signed/unsigned mismatch
// C5045 - Compiler will insert Spectre mitigation for memory load if /Qspectre switch specified
// C4263 - 'void ApiVulkanSample::render(float)': member function does not override any base class virtual member function
// C4264 - 'void vkb::VulkanSample::render(vkb::CommandBuffer &)': no override available for virtual member function from base 'vkb::VulkanSample'; function is hidden
// C4389 - signed/unsigned mismatch
// C4868 - compiler may not enforce left-to-right evaluation order in braced initializer list
// C4623 - 'vkb::sg::AnimationChannel': default constructor was implicitly defined as deleted
// C4242 C4244 -  conversion from 'X' to 'Y', possible loss of data (occurs in algorithm annoyingly...)
// C5039 - pointer or reference to potentially throwing function passed to 'extern "C"' function under -EHc. Undefined behavior may occur if this function throws an exception (GLFW window callbacks)

// Concerns C4263, C4264, C5039
```

I disabled all these warnings to get the build to work. Some of the smaller things I fixed but these are the main ones that popped up. I think some of these are excessive (C4820, C4263, C4264...) and so would like further opinions before going through and fixing them. MSVC in general seems to include a lot of warnings at this level which seem to cause more issues than they solve...

Also some of these warnings occur in the standard library which is out of our control to fix. A solution to avoiding all of this is to wrap both third parties and the standard library with the enable/disable warnings macros.

TomAtkinsonArm committed:
initial MSVC investigation

gpx1000 committed:
Merge branch 'master-upstream' into warning_removal_pr-320

# Conflicts:
#	app/vulkan_samples.cpp
#	framework/api_vulkan_sample.cpp
#	framework/common/strings.cpp
#	framework/core/device.cpp
#	framework/core/instance.cpp
#	framework/platform/application.cpp
#	framework/platform/application.h
#	framework/platform/glfw_window.cpp
#	framework/platform/input_events.h
#	framework/platform/parser.h
#	framework/platform/parsers/CLI11.cpp
#	framework/platform/parsers/CLI11.h
#	framework/platform/platform.cpp
#	framework/platform/platform.h
#	framework/platform/unix/unix_platform.h
#	framework/platform/window.cpp
#	framework/platform/window.h
#	samples/api/compute_nbody/compute_nbody.cpp
#	samples/performance/msaa/msaa.cpp

gpx1000 committed:
remove warnings and enable clang/gcc to build warning free.

TomAtkinsonArm committed:
Merge branch 'warning_removal_pr-320' of https://github.com/gpx1000/Vulkan-Samples into gpx1000-warning_removal_pr-320

TomAtkinsonArm committed:
Merge branch 'gpx1000-warning_removal_pr-320' into more-warnings

TomAtkinsonArm committed:
CLI patch

TomAtkinsonArm commented:
depreciated. Framework v2.0 uses higher warning levels
